scripts: west: commands: build: add --shield argument for specifying shields

henrikbrixandersen opened the pull request:
This patch series adds support for doing `west build -b ... --shield foo` instead of having to do `west build -b ... -- -DSHIELD=foo`. It contains the following additions:
- scripts: west: build: add snippet argument to usage output (missing, identified while adding support for --shield).
- scripts: west: completion: bash: add snippet arguments to build command (missing, identified while adding support for --shield).
- scripts: list_shields: add script for listing the supported shields.
- scripts: west: commands: shields: add cmd for listing supported shields.
- scripts: west: commands: completion: bash: add "west shields" completion.
- scripts: west: commands: build: add arguments for specifying shields.
- scripts: west: commands: completion: bash: support shield arguments.

If this is accepted, a couple of follow-up PRs could:
- Refactor `cmake/modules/shields.cmake` to use `scripts/list_shields.py`.
- Change `doc/_extensions/zephyr/application.py` to recommend using `--shield|-i` in generated west command documentation.

henrikbrixandersen committed:
scripts: west: build: add snippet argument to usage output

Add example snippet argument to usage output of the "west build" command
and use the singular "SNIPPET" in the generated help text.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen committed:
scripts: west: completion: bash: add snippet arguments to build command

Add snippet arguments to the "west build" command completion.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen committed:
scripts: list_shields: add script for listing the supported shields

Add Python script for listing the supported shields.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen committed:
scripts: west: commands: shields: add cmd for listing supported shields

Add "west shields" command for listing supported shields.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen committed:
scripts: west: commands: completion: bash: add "west shields" completion

Add Bash completion for the "west shields" command.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen committed:
scripts: west: commands: build: add argument for specifying shields

Add "west build" argument for specifying shields. This saves users from
having to do "west build ... -- -DSHIELD=...", making shields a much better
integrated feature.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen committed:
scripts: west: commands: completion: bash: support shield argument

Add Bash completion support for "west build" shield argument.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

fabiobaltieri commented on scripts/west_commands/build.py:
```diff
@@ -24,6 +24,7 @@
 
 BUILD_USAGE = '''\
 west build [-h] [-b BOARD[@REV]]] [-d BUILD_DIR]
+           [-s SNIPPET] [-i SHIELD]
```
isn't snippet uppercase S? then you could take the lowercase one for this

henrikbrixandersen replied:
Yeah, `-s` is used for source directory for backwards compatibility. I've fixed the capitalisation of the `-S`, thanks.

henrikbrixandersen committed:
scripts: west: commands: build: fix pylint issue

Do not end except block with a useless return, use a pass instead. This
fixes pylint R1711: "Useless return at end of function or method
(useless-return)".

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

fabiobaltieri reviewed (dismissed):
-i is a bit sad but I guess there's only so many letters...

carlescufi reviewed (dismissed):
Looks great!

tejlmand commented on scripts/west_commands/build.py:
```diff
@@ -24,6 +24,7 @@
 
 BUILD_USAGE = '''\
 west build [-h] [-b BOARD[@REV]]] [-d BUILD_DIR]
+           [-S SNIPPET]
```
Thanks for noticing, but it's plural in the `west help build` output because the dest variable is in plural:
```
$ west help build
usage: west build [-h] [-b BOARD[@REV]]] [-d BUILD_DIR]
...
   -S SNIPPETS, --snippet SNIPPETS
                        add the argument to SNIPPET; may be given multiple times. Forces CMake to run again if given. Do not use this option with manually specified -DSNIPPET... cmake arguments: the results are undefined

```

Ref: https://github.com/zephyrproject-rtos/zephyr/blob/3412e55d183ea098f7420da485d76902bafbc05f/scripts/west_commands/build.py#L139


I have no preference regarding singular or plural, only regarding consistency :)

Wonder if we can get the detailed help text to be singular, so they match.

henrikbrixandersen replied:
Good catch, thanks. I have added `metavar` arguments to use singular form for both SNIPPET and SHIELD.

tejlmand commented on scripts/list_shields.py:
```diff
@@ -0,0 +1,71 @@
+#!/usr/bin/env python3
```
Non blocking comment.

The long term goal is to get rid of the `list_boards.py` and unify this under [list_hardware.py](https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/list_hardware.py) because listing of boards now requires sourcing of the SoC for the CPU cluster knowledge (and in future maybe SiP in addition).

Reason for keeping the `list_boards` short term is that we still support HWMv1.

So although `list_shields.py` is small, I think it make sense if this can be done as part of list_hardware.

Of course it can also be reasoned that because shields has not been touched as part of the new hw model and therefore is still looked up using `Kconfig.shield` + `<name>.overlay`  then keeping it independent from list_hardware makes sense.

henrikbrixandersen replied:
Yeah, I was a bit confused as to the purpose of the new `list_hardware.py` script. I initially tried integrating shield support there, but it seemed like it was more geared towards SoCs and architectures, whereas `list_boards.py` seems closed to what I wanted to add.

I'd prefer to leave `list_shields.py`.

tejlmand commented on scripts/west_commands/build.py:
```diff
@@ -143,6 +143,13 @@ def do_add_parser(self, parser_adder):
                            Do not use this option with manually specified
                            -DSNIPPET... cmake arguments: the results are
                            undefined''')
+        group.add_argument('-i', '--shield', dest='shields',
```
we don't need a short argument variant for all arguments.

And especially in this case I'm not fond of reserving `-i` for something like shield.
It could happen in future that a long arg starting with `i` is introduced and therefor making it natural to use `-i` for such argument.

henrikbrixandersen replied:
That's fair. I wasn't too happy with `-i` myself. I have removed it, leaving only the long option.

tejlmand commented on scripts/west_commands/completion/west-completion.bash:
```diff
@@ -754,6 +760,10 @@ __comp_west_build()
 			__set_comp_west_boards
 			return
 			;;
+		--shield|-i)
```
please update zsh and fish in addition.
https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/west_commands/completion/west-completion.fish
https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/west_commands/completion/west-completion.zsh

henrikbrixandersen replied:
I don't use either of those shells, sorry.

tejlmand replied:
a bit unfortunate, I use the `fish` shell so can take a look in future.

In the meantime, can you create a follow-up enhancement issue, so that at least we have some visibility that completion of shields are missing for zsh and fish ?

tejlmand reviewed (changes requested):
really not fond of reserving `-i` for this.

Why not just keep the long arg `--shield`, afaict that still an improvement to users.

tejlmand reviewed (approved):
Approved, please create enhancement issues for the missing shell completions

henrikbrixandersen commented:
> > Approved, please create enhancement issues for the missing shell completions
> 
> I'll create enhancement issues for that once this is merged. Thanks.

@tejlmand I have created the following enhancement requests as agreed:
- https://github.com/zephyrproject-rtos/zephyr/issues/73128
- https://github.com/zephyrproject-rtos/zephyr/issues/73129

kartben commented:
@henrikbrixandersen another useful follow-up would probably be to update all the READMEs of the shields to now point to this much more convenient way of enabling shields. `zephyr-app-commands` Sphinx extension also needs updating so that the output also uses `--shield` when a shield is specified. If you find time to open the issue(s), I am happy to tackle the associated PRs :)

henrikbrixandersen commented:
> @henrikbrixandersen another useful follow-up would probably be to update all the READMEs of the shields to now point to this much more convenient way of enabling shields. `zephyr-app-commands` Sphinx extension also needs updating so that the output also uses `--shield` when a shield is specified. If you find time to open the issue(s), I am happy to tackle the associated PRs :) 

Thank you for the reminder! I'll open improvement PRs.

henrikbrixandersen commented:
> > @henrikbrixandersen another useful follow-up would probably be to update all the READMEs of the shields to now point to this much more convenient way of enabling shields. `zephyr-app-commands` Sphinx extension also needs updating so that the output also uses `--shield` when a shield is specified. If you find time to open the issue(s), I am happy to tackle the associated PRs :)
> 
> Thank you for the reminder! I'll open improvement PRs.

Completely forgot about this until today, but here it is: https://github.com/zephyrproject-rtos/zephyr/pull/75311
